### PR TITLE
Vagrant module

### PR DESCRIPTION
**Pull request by vernondcole**
### What does this PR do?

Adds a module to control Vagrant virtual machines on a minion.

This follows the same outline and uses the same high-level control methods  as the `virt` module. It contains far fewer methods, however, since it operates on a much less detailed level.

### What issues does this PR fix or reference?

This begins a different implementation of #43298 as requested.

### New Behavior

`salt my_laptop vagrant.init x1 cwd=/my_dir start=True` will start a Vagrant VM

Following the example of `virt`, the user must do `vagrant.init` before doing `vagrant.start`. A `start=` keyword is also present with the default set to `False`. 

Commands `destroy`, `start`, `reboot`, `pause`, `stop`, and `shutdown` are present as in the `virt` module. In this module, `stop` and `shutdown` are synonyms.  

The command (actually query) `list_domains` is retained from `virt`, even though the name is inaccurate for Vagrant VM's.  It lists all VM's on the host machine, regardless of ownership.  It does this using cached information, and thus may be inaccurate.

The queries `list_active_vms`, `list_inactive_vms` and `vm_state` are copied from the `virt` module, but in this module the user must provide a `cwd=` directory (and possibly a `runas=` username) to query the current running state of the VMs.

The queries `vagrant.version` and `vagrant.get_ssh_config` are added.  The latter will be used to get the information that the salt-cloud driver will need to deploy the VM.

### Tests written?

Yes -- but they are unit tests for non-important functions only.  They don't really test whether virtual machines can be controlled.

### CODE REVIEW

**Comment by vernondcole** [comment]
@gtmanfred You might want to review this, too.

File: salt/modules/vagrant.py
Comment (gtmanfred) [inline salt/modules/vagrant.py]: Is there a reason to use `subprocess.call` instead of `__salt__['cmd.shell']`?

The other benefit is that in 2017.7, runas is a state runtime variable, so that you don't have to provide runas in the state if you write a vargant state as well, the cmd.shell will just get the runas user from the `__context__` object.

And then you also do not need the _runas_sudo command, you can just pass which user to runas to the cmd module.

**Review by gtmanfred — CHANGES_REQUESTED** [review]
This is a good start for this, I just have the one major concern that I would like to see cmd.shell used instead of subprocess python module.

File: salt/modules/vagrant.py
Comment (vernondcole) [reply salt/modules/vagrant.py]: I did not even think about using __salt__['cmd.shell'].  Too used to doing it all myself. I'll look at that.

Yes, I am planning to write a vagrant state module, just need this as a starting point.

Comment (vernondcole) [reply salt/modules/vagrant.py]: Great suggestion!  The module is now 70 lines shorter.

**Comment by vernondcole** [comment]
Conversion from subprocess to cmd.shell and cmd.retcode simplified the module code by about 70 lines, in addition to adding features.

**Review by gtmanfred — APPROVED** [review]

File: salt/modules/vagrant.py
Comment (gtmanfred) [inline salt/modules/vagrant.py]: @cachedout does this need to be `{0}` still for following the old standards, or can we stop doing that now that we have dropped py2.6?

**Comment by cachedout** [comment]
@vernondcole There is a failing test here that needs to be addressed, please:

https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu14-n/14881/testReport/junit/unit.modules.test_vagrant/VagrantTestCase/test_vagrant_destroy_removes_cached_entry/

**Comment by vernondcole** [comment]
The remaining test failures are not related to my PR.

File: salt/modules/vagrant.py
Comment (cachedout) [inline salt/modules/vagrant.py]: IMHO, this would be cleaner if we just did `if 'machine' not in vm_:`

File: salt/modules/vagrant.py
Comment (cachedout) [inline salt/modules/vagrant.py]: We should not ever call `print` directly from execution modules. Was this just for debugging/development?

File: salt/modules/vagrant.py
Comment (cachedout) [inline salt/modules/vagrant.py]: Same here. Please remove this `print`

File: salt/modules/vagrant.py
Comment (cachedout) [inline salt/modules/vagrant.py]: Another `print` to remove

File: salt/modules/vagrant.py
Comment (cachedout) [inline salt/modules/vagrant.py]: We could just do a `__func_alias__` here, I think. See other modules for an example.

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/modules/vagrant.py
Comment (vernondcole) [reply salt/modules/vagrant.py]: good catch.  I usually mark my debug lines with a triple hash, like print(xxx) ###,  but missed this time and then pasted it twice more.

File: salt/modules/vagrant.py
Comment (vernondcole) [reply salt/modules/vagrant.py]: That's a nifty trick! I was unaware of it.

I tried the suggestion.  I converted the signature to `start_(machine, vm_=None)` and then filled in the vm_ when it was null. 

After `make html`, the resulting html document contains the reference to the optional `vm_` argument.

That argument is useful for internal calls only.  I could not figure out a good way to explain that in the docs, so I reverted the code to maintain the separation between the external and internal calls.  I think the cleaner documentation is worth the cost of the extra call in this case.

**Comment by vernondcole** [comment]
@cachedout @gtmanfred I am worried that this PR is getting too stale.  I have another round of changes from lessons I learned while making the Vagrant state module.  Should I submit them here now, or wait and include them with the PR for the state?

**Comment by rallytime** [comment]
@vernondcole Thanks for submitting your changes. I'll get @gtmanfred and @cachedout to do another round of reviews.

If you could wait to submit your changes in another PR that would be best so we can keep the relevant changes reviewed, then review your next changes.

File: salt/modules/vagrant.py
Comment (gtmanfred) [inline salt/modules/vagrant.py]: Why are we keeping the cache for 10 years?

This should be able to be configured, and it looks like elsewhere it defaults to `10`

**Review by gtmanfred — CHANGES_REQUESTED** [review]

File: salt/modules/vagrant.py
Comment (vernondcole) [reply salt/modules/vagrant.py]: I am purposely misusing cache as a database. (Somebody suggested that I should do that in the saltify cloud driver.)  

The cache stores the mapping between the Salt node name and the Vagrant machine name (and directory). So when the user types "vagrant.destroy x1" I can determine which vagrant client machine she means. It needs to be maintained forever, and ten years seems like a nice approximation of that.

Comment (vernondcole) [reply salt/modules/vagrant.py]: The other option would be to use a real database.  We could use sqlite, but this felt like it fit the application better.

Comment (gtmanfred) [reply salt/modules/vagrant.py]: This should use a real database then.  Sqlite is probably fine.

If you wanted it to be pluggable you can use `sdb` https://docs.saltstack.com/en/latest/topics/sdb/

**Comment by vernondcole** [comment]
@cachedout @gtmanfred @rallytime 
The changes requested were so extensive that I merged in my updated version in before re-designing the storage system.  This should merge back in to my vagrant-state branch smoothly that way.

This storage method does smell better.  I mixed both types of record (name and machine) in one table by making sure the keys would not collide.  The only funny thing was a kludge to cover for the fact that the sdb SQLite driver has no .delete() method. I can't fathom why.  So I try to .delete, and when it fails (I am hoping that the driver will be improved some day) I replace the meaningful value for that key with an empty value.  It's ugly, but it works.

File: salt/modules/vagrant.py
Comment (gtmanfred) [inline salt/modules/vagrant.py]: I am ok with this if everyone else is.

**Review by gtmanfred — APPROVED** [review]

**Comment by vernondcole** [comment]
Dammit!  I have been bitten by git again!  I had (I think) one line to change to make the tests pass:
I was writing my SQLite file outside of /tmp/salt-tests-tmpdir/ which I suspect is a sandbox.  I will try to rebase things to remove all of this mess...

**Comment by vernondcole** [comment]
Okay, I think it's back to sanity again.

**Comment by cachedout** [comment]
Unfortunetely things did still not go well under Python 3: https://jenkins.saltstack.com/job/PR/job/salt-pr-linode-ubuntu16-py3/2354/

**Comment by cachedout** [comment]
The linter also detected an issue which needs to be addressed, please: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/15087/violations/file/salt/modules/vagrant.py/

**Comment by vernondcole** [comment]
@cachedout The Python3 error is not my bug. It is a  bug in salt.utils.sdb for which no unit tests were written.

I have submitted issue #43815 which supplies the missing unit test.

File: salt/modules/vagrant.py
Comment (gtmanfred) [inline salt/modules/vagrant.py]: can you use `__utils__['sdb.sdb_set']` instead of importing the salt.utils.sdb module?

same for the rest of the places you are using salt.utils.

**Review by gtmanfred — CHANGES_REQUESTED** [review]

File: salt/modules/vagrant.py
Comment (vernondcole) [reply salt/modules/vagrant.py]: Yes, I will revert that.  I thought that the lazy loader was causing my Python3 errors, but I was wrong.

**Comment by vernondcole** [comment]
@cachedout @gtmanfred I think this is it, guys.  All the failures are known problems unrelated to this PR.

**Comment by cachedout** [comment]
re-run lint

**Comment by rallytime** [comment]
re-run py3

**Comment by vernondcole** [comment]
@cachedout @rallytime @gtmanfred 
This should take care of the Python3 problem once and for all.  It turns out that my unit tests were actually integration tests by Salt definitions.  So I have completely rewritten my unit tests using Mock and patch. Since they no longer call external modules, there will be no trouble with funky loader bugs.

**Comment by vernondcole** [comment]
Question:  should I rework my old version of the unit test and submit it as an integration test? (In a different PR?)

**Comment by gtmanfred** [comment]
You still have one pylint error

```
00:03:57 tests/unit/modules/test_vagrant.py:32: [E8301(expected-1-blank-line-found-0), ] PEP8 E301: expected 1 blank line, found 0

```

**Review by gtmanfred — APPROVED** [review]

**Comment by vernondcole** [comment]
@cachedout Will this do?

**Comment by garethgreenaway** [comment]
@vernondcole We just added codeclimate in recently and we're still tuning the settings.

**Comment by vernondcole** [comment]
I can't really know whether my one-line code cleanup actually made the difference, but a large number of vagrant-associated warnings went away, and the Python3 jenkins job now has a green check mark.

**Comment by garethgreenaway** [comment]
re-run ubuntu

**Comment by garethgreenaway** [comment]
re-run ubuntu

**Comment by vernondcole** [comment]
@rallytime @garethgreenaway 
Are we waiting for approval from @cachedout ?

**Comment by garethgreenaway** [comment]
Waiting for tests to pass and for it to be in sync with the parent branch. Hopefully this current run will complete successfully.

**Comment by garethgreenaway** [comment]
These failures are unrelated and I saw all tests pass earlier today, branch was out of sync with develop so it was updated.

**Pull request merged**
